velox/type/Type.h

```diff
@@ -569,9 +579,16 @@ class Type : public Tree<const TypePtr>, public velox::ISerializable {
 template <TypeKind KIND>
 class TypeBase : public Type {
  public:
-  using NativeType = TypeTraits<KIND>;
-
-  TypeBase() : Type{KIND} {}
+  using NativeType = typename TypeTraits<KIND>::NativeType;
+
+  explicit TypeBase(bool providesCustomComparison = false)
+      : Type{KIND, providesCustomComparison} {
+    if (providesCustomComparison) {
+      VELOX_CHECK(
+          TypeTraits<KIND>::isPrimitiveType && TypeTraits<KIND>::isFixedWidth,
```

Comment:
Unfortunately Type doesn't have easy access to the properties of the TypeKind (it's passed by value), which is why I put it in the TypeBase constructor.  I can call those functions though (I initially avoided it because they're virtual, but according to the spec that means they'll use the local definition which is equivalent).